service/src/test/java/bio/terra/workspace/service/admin/AdminServiceTest.java

```diff
@@ -0,0 +1,231 @@
+  @Autowired WorkspaceConnectedTestUtils connectedTestUtils;
+  @Autowired UserAccessUtils userAccessUtils;
+  @Autowired CrlService crlService;
+  @Autowired WorkspaceActivityLogDao workspaceActivityLogDao;
+
+  private IamCow iamCow;
+  List<UUID> workspaceIds = new ArrayList<>();
+  List<String> projectIds;
+
+  @BeforeEach
+  void setup() {
+    iamCow = crlService.getIamCow();
+    workspaceIds.add(
+        connectedTestUtils
+            .createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest())
+            .getWorkspaceId());
+    workspaceIds.add(
+        connectedTestUtils
+            .createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest())
+            .getWorkspaceId());
+    workspaceIds.add(
+        connectedTestUtils
+            .createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest())
+            .getWorkspaceId());
+    projectIds =
+        workspaceDao.getWorkspaceToCloudContextMap(CloudPlatform.GCP).values().stream()
+            .map(cloudContext -> GcpCloudContext.deserialize(cloudContext).getGcpProjectId())
+            .toList();
+    // The existing project has incomplete permissions on PROJECT_READER
+    for (String project : projectIds) {
+      updateCustomRole(INCOMPLETE_PROJECT_READER, project);
+    }
+  }
+
+  @AfterEach
+  void cleanUp() {
+    jobService.setFlightDebugInfoForTest(null);
+    for (UUID workspaceId : workspaceIds) {
+      connectedTestUtils.deleteWorkspaceAndGcpContext(
+          userAccessUtils.defaultUserAuthRequest(), workspaceId);
+    }
+  }
+
+  @Test
+  public void syncIamRole_newPermissionsAddedToCustomRoleProjectReader() {
+    OffsetDateTime lastChangeTimestampOfWorkspace1 =
+        workspaceActivityLogDao.getLastUpdateDetails(workspaceIds.get(0)).get().getChangeDate();
+    OffsetDateTime lastChangeTimestampOfWorkspace2 =
+        workspaceActivityLogDao.getLastUpdateDetails(workspaceIds.get(1)).get().getChangeDate();
+    OffsetDateTime lastChangeTimestampOfWorkspace3 =
+        workspaceActivityLogDao.getLastUpdateDetails(workspaceIds.get(2)).get().getChangeDate();
+    // Test idempotency of steps by retrying them once.
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(
+        RetrieveGcpIamCustomRoleStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(GcpIamCustomRolePatchStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().doStepFailures(retrySteps).build());
+
+    String jobId =
+        adminService.syncIamRoleForAllGcpProjects(
+            userAccessUtils.defaultUserAuthRequest(), /*wetRun=*/ true);
+    jobService.waitForJob(jobId);
+    for (String projectId : projectIds) {
+      assertProjectReaderRoleIsUpdated(
+          projectId, CUSTOM_GCP_PROJECT_IAM_ROLES.get(WsmIamRole.READER).getIncludedPermissions());
+    }
+    assertTrue(
```

Comment:
Should there be a test for:

- No new permissions, workspace is not updated
- No activity log

Maybe add to this test

Reply:
done